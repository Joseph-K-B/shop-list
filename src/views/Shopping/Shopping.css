.shop {
    display: grid;
    max-width: 100vw;
    grid-template-columns: 1fr 3fr;
    text-align: center;
}

.add {
    display: flex;
    flex-direction: column;
    justify-content: space-between;
    max-height: 10rem;
    max-width: 10rem;
    border: 0.25rem black solid;
    background-color: lawngreen;
    padding: 5%;
}

input {
    max-width: 8rem;
}