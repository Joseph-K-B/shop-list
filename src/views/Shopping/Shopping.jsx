import { useReducer } from "react"
import ListControls from "../../components/controls/ListControls";
import ShopList from "../../components/list/ShopList";
import itemReducer from "../../utils/ItemReducer"

const initialState = [
  {id: 1, name: 'test', done: false},
  {id: 2, name: 'test-2', done: false},
]

function Shopping() {
    const [items, dispatch] = useReducer(itemReducer, initialState);
    
    const handleAdded = (name) => {
      dispatch({
        type: 'added',
        id: items.length + 1,
        name,
      });
    };

    const handleUpdate = (item) => 
    dispatch({
      type: 'changed',
      item
    });

    const handleDelete = (id) => {
      dispatch({
        type: 'deleted',
        id
      });
    };

    return(
      <>
        <h1>Shopping List</h1>
        <ListControls onAdded={handleAdded} />
        <ShopList items={items} onUpdate={handleUpdate} onDeleteItem={handleDelete}/>
      </>
    )
}

export default Shopping