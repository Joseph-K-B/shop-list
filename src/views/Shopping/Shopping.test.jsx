import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import Shopping from './Shopping';

it('allows user to add, update, and delete items', async () => {
    const {container} = render(
        <Shopping items={[{id: 1, name: 'test', done: false}]}/>
    );

    const nameInput = screen.getByLabelText(/new item input/i);
    const submitBtn = screen.getByLabelText(/submit new item/i);


    userEvent.type(nameInput, 'name');
    userEvent.click(submitBtn);

    const deleteBtn = await screen.findByLabelText(/open delete form/i);
    const editBtn = await screen.findByLabelText(/open update form/i);
    const newItem = await screen.findByLabelText(/list-item/i);
    
    expect(editBtn).toBeInTheDocument();

    userEvent.click(editBtn);
    const updateInput = await screen.findByLabelText(/update item input/i);
    const updateBtn = await screen.findByLabelText(/apply update/i);

    expect(updateInput).toBeInTheDocument();
    expect(updateBtn).toBeInTheDocument();
    
    userEvent.type(updateInput, 'better-name');
    userEvent.click(updateBtn);

    expect(deleteBtn).toBeInTheDocument();
    expect(nameInput).toBeInTheDocument();
    expect(submitBtn).toBeInTheDocument();
    expect(newItem).toBeInTheDocument();
    expect(container).toMatchSnapshot();
});
