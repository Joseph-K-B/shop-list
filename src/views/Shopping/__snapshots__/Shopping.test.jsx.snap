// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`allows user to add, update, and delete items 1`] = `
<div>
  <h1>
    Shopping List
  </h1>
  <form>
    <input
      aria-label="new item input"
      type="text"
      value=""
    />
    <button
      aria-label="submit new item"
      type="submit"
    >
      Add Item
    </button>
  </form>
  <ul>
    <li>
      <section
        aria-label="list-item"
      >
        <h3>
          namebetter-name
        </h3>
        <input
          type="checkbox"
          value="false"
        />
        <button
          aria-label="open delete form"
          type="button"
        >
          Remove
        </button>
        <button
          aria-label="open update form"
        >
          Edit
        </button>
      </section>
    </li>
  </ul>
</div>
`;
